In [38]:
import nltk
from nltk import FreqDist, ConditionalFreqDist, word_tokenize
from nltk.corpus import stopwords
import matplotlib.pyplot as plt
%matplotlib inline
import os
from sklearn.feature_extraction.text import TfidfVectorizer

In [39]:
os.chdir('/Users/racheldyap/Desktop/DataScienceFoundations/Session_7')

In [40]:
songs = []

for file in os.listdir('/Users/racheldyap/Desktop/DataScienceFoundations/Session_7'):
    if file.endswith('.txt'):
        with open(str(file)) as f:
            songs.append(f.read().replace('\n', ' '))

In [56]:
tokenize = lambda doc: doc.lower().split(" ")

In [57]:
sklearn_tfidf = TfidfVectorizer(norm='l2',min_df=0, use_idf=True, smooth_idf=False, sublinear_tf=True, tokenizer=tokenize)

In [58]:
sklearn_representation = sklearn_tfidf.fit_transform(songs)

In [59]:
def cosine_similarity(vector1, vector2):
    dot_product = sum(p*q for p,q in zip(vector1, vector2))
    magnitude = math.sqrt(sum([val**2 for val in vector1])) * math.sqrt(sum([val**2 for val in vector2]))
    if not magnitude:
        return 0
    return dot_product/magnitude

In [60]:
import math

for i in range(7):
    for y in range(i+1,7):
        print(i,y, cosine_similarity(sklearn_representation.toarray()[i], sklearn_representation.toarray()[y]))

0 1 0.159722344622
0 2 0.245128300669
0 3 0.164376215557
0 4 0.210425398329
0 5 0.21764296934
0 6 0.0791437902704
1 2 0.198654045995
1 3 0.14274587974
1 4 0.144152166451
1 5 0.126074999207
1 6 0.0906362810996
2 3 0.185592604352
2 4 0.229413325808
2 5 0.189575826027
2 6 0.100041686064
3 4 0.163513105209
3 5 0.159766833781
3 6 0.074116145503
4 5 0.180645285235
4 6 0.121154178333
5 6 0.100021053946


#### most similar song are 7 Days (0) and Fill Me in (2) with score of 0.245

In [66]:
songs_tokenize = [tokenize(d) for d in songs]

# remove stopwords from tokenize songs
porter = nltk.PorterStemmer()
all_songs = []

for lyrics in songs_tokenize:
    lyrics_stopwords = [t for t in lyrics if t not in stopwords.words('english')]
    lyrics_stemmed = [porter.stem(t) for t in lyrics_stopwords]
    all_songs.append(lyrics_stemmed)

In [67]:
all_lyrics = [' '.join(lyric) for lyric in all_songs]

In [73]:
sklearn_representation = sklearn_tfidf.fit_transform(all_lyrics)

In [77]:
for i in range(7):
    for y in range(i+1,7):
        print(i,y, cosine_similarity(sklearn_representation.toarray()[i], sklearn_representation.toarray()[y]))

0 1 0.104030397477
0 2 0.173627920786
0 3 0.0767646796954
0 4 0.140389619075
0 5 0.175178920102
0 6 0.035891256556
1 2 0.137850845198
1 3 0.0664567549833
1 4 0.084919916086
1 5 0.0596876627262
1 6 0.0475010774226
2 3 0.0898675416479
2 4 0.144144669987
2 5 0.132749012632
2 6 0.0624504970729
3 4 0.0937929001491
3 5 0.0580699154989
3 6 0.0237930703818
4 5 0.10955847793
4 6 0.0823834856376
5 6 0.0520853437211


#### after removing the stopwords, the results have changed.
most similar words are now 7 Days (#0) and Rendezvous (#5)

In [84]:
from nltk.corpus import treebank, stopwords

In [86]:
cfd = ConditionalFreqDist()
stopwords_list = stopwords.words('english')

def is_noun(tag): # we define a noun; noun is tagged as those below in red
    return tag.lower() in ['nn','nns','nn$','nn-tl','nn+bez', 'nn+hvz', 
                           'nns$','np','np$','np+bez','nps', 'nps$','nr',
                           'np-tl','nrs','nr$']

In [88]:
for sentence in treebank.tagged_sents():
    for (index, tagtuple) in enumerate(sentence):
        (token, tag) = tagtuple
        token = token.lower()
        if token not in stopwords_list and is_noun(tag): 
            window = sentence[index+1:index+5] 
            for (window_token, window_tag) in window:
                window_token = window_token.lower()
                if window_token not in stopwords_list and is_noun(window_tag):
                    cfd[token][window_token] += 1

In [90]:
print(cfd['chairman'].max())

chief


In [91]:
print(cfd['profit'].max())

%


In [92]:
print(cfd['president'].max())

chief


In [93]:
print(cfd['investors'].max())

bonds


In [94]:
print(cfd['city'].max())

streets


In [96]:
print(cfd['chief'].max())

officer


#### chief seems to be most associated with any position that holds authority